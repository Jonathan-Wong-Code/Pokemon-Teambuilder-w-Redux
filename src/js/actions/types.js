export const ADD_TEAM = 'ADD_TEAM';
export const DELETE_TEAM = 'DELETE_TEAM';
export const EDIT_TEAM = 'EDIT_TEAM';
export const SET_SAVED_TEAMS = 'SET_SAVED_TEAMS';
export const RESET_SAVED_TEAMS = 'RESET_SAVED_TEAMS';
export const LOGIN = 'LOGIN';
export const LOGOUT = 'LOGOUT';
export const LOGIN_GUEST = 'LOGIN_GUEST';
