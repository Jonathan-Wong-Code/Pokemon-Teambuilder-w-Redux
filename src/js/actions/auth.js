import database, { firebase, googleAuthProvider } from './../firebase/firebase';
import { LOGIN, LOGOUT, LOGIN_GUEST, SET_SAVED_TEAMS } from './types';
import regeneratorRuntime from 'regenerator-runtime';

export const startLogin = () => async dispatch => {
  await firebase.auth().signInWithPopup(googleAuthProvider);
};

export const startLogout = () => async dispatch => {
  await firebase.auth().signOut();
};

export const login = (uid) => ({
  type : LOGIN,
  uid
});

export const loginGuest = () => async (dispatch, getState) => {
  dispatch({ type : LOGIN_GUEST }); 

  const uid = getState().auth.uid;
  const response = await database.ref(`users/${uid}/pokemon`).once('value');
  const savedTeams = {};
  response.forEach(team => {
    savedTeams[team.key] = {
      ...team.val(),
      id : team.key
    }
  });

  dispatch({
    type : SET_SAVED_TEAMS,
    savedTeams
  });
};

export const logout = () => ({
  type : LOGOUT
});
