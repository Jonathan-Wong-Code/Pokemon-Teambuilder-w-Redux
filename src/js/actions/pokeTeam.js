import { ADD_TEAM, DELETE_TEAM, EDIT_TEAM, SET_SAVED_TEAMS } from './types';
import database from './../firebase/firebase';
import regeneratorRuntime from "regenerator-runtime";

export const addPokeTeam = (pokeTeam) => async (dispatch, getState) => {
  const uid = getState().auth.uid;
  const response = await database.ref(`users/${uid}/pokemon`).push(pokeTeam);
  const newTeamObject = ({
    ...pokeTeam,
    id: response.key
  })

  dispatch({
    type : ADD_TEAM,
    team : newTeamObject
  });

  return response.key
};

export const deletePokeTeam = (id) => async (dispatch, getState) => {
  const uid = getState().auth.uid;
  database.ref(`users/${uid}/pokemon/${id}`).remove();

  dispatch({
    type : DELETE_TEAM,
    id
  });
};

export const editPokeTeam = (pokeTeam) => async (dispatch, getState) => {
  const uid = getState().auth.uid;
  database.ref(`users/${uid}/pokemon/${pokeTeam.id}`).update(pokeTeam);

  dispatch({
    type : EDIT_TEAM,
    team : pokeTeam
  });
};

export const setSavedTeams = () => async (dispatch, getState) => {
  const uid = getState().auth.uid;
  const response = await database.ref(`users/${uid}/pokemon`).once('value');
  const savedTeams = {};
  response.forEach(team => {
    savedTeams[team.key] = {
      ...team.val(),
      id : team.key
    }
  });

  dispatch({
    type: SET_SAVED_TEAMS,
    savedTeams
  });
}
