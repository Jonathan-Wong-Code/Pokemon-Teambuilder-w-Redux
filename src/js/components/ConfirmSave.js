import React from 'react';

export class ConfirmSave extends React.Component {
  constructor() {
    super();

    this.closeBtn = React.createRef();
  }

  componentDidMount() {
    this.closeBtn.current.focus();
  }
  
  render() {
    return (
      <div 
        className='modal__background' 
        onClick={this.props.handleToggleSaveModal} 
        data-test='confirm-save-component'
      >
        <div className='confirm-save__modal' onClick={(e) => e.stopPropagation()}
          data-test='confirm-save-body'
        >
          <p data-test='confirm-save-text'>Saved <span className='confirm-save__name'>
              {this.props.pokemon.name}
            </span>
          </p>
          <button 
            onClick={this.props.handleToggleSaveModal} 
            className='btn confirm-save__btn'
            data-test='confirm-save-close-btn'
            ref={this.closeBtn}
          >
          Close
          </button>
        </div>
      </div>
    );
  }
}

export default ConfirmSave;
