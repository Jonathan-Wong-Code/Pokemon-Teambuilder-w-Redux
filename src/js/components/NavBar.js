import React from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout, logout } from './../actions/auth';

export const NavBar = ({ startLogout, history, auth, logout }) => {
  const onLogoutClick = () => {
    if (auth === 'guest') {
      logout();
      history.push('/');
    } else {
      startLogout();
    }
  };

  return (
    <nav className='navbar' data-test='nav-bar-component'>
      <ul className='navbar__nav'>
        <li className='navbar__item'>
          <NavLink to='/create' activeClassName='is-active' className='navbar__link btn'>
            Create Team
          </NavLink>
        </li>
        <li className='navbar__item'>
          <NavLink to='/SavedTeams' activeClassName='is-active' className='navbar__link btn'>
            Saved Teams
          </NavLink>
        </li>
        <li className='navbar__item'>
          <button 
            onClick={onLogoutClick} 
            className='btn navbar__logout'  
            data-test='nav-bar-logout'>
            Logout
          </button>
        </li>
      </ul>
    </nav>
  );  
};

const mapStateToProps = (state) => ({
  auth : state.auth.uid
});

const mapDispatchToProps = (dispatch) => ({
  startLogout : () => dispatch(startLogout()),
  logout : () => dispatch(logout())
});

export default connect(mapStateToProps, mapDispatchToProps)(NavBar);
