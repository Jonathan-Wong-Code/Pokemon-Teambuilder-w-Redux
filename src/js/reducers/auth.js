import { LOGIN, LOGIN_GUEST, LOGOUT } from './../actions/types';

const initialAuthState = {
  uid : null
}

const authReducer = (state = initialAuthState, action) => {
  switch (action.type) {

    case LOGIN:
      return {
        uid : action.uid
      };
    
    case LOGIN_GUEST:
      return {
        uid : 'guest'
      };

    case LOGOUT:
      return {
        uid : null
      };

    default:
      return state;
  }
};

export default authReducer;
