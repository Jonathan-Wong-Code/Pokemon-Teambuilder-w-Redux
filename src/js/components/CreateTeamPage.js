import React from 'react';
import { connect } from 'react-redux';
import BuildPokeTeamPage from './BuildPokeTeamPage';
import { addPokeTeam } from './../actions/pokeTeam';

export const CreateTeamPage = ({ history, addPokeTeam }) => {
  const handleAddTeam = (pokeTeam) => {
    addPokeTeam(pokeTeam);
  };
 
  return (
    <BuildPokeTeamPage 
      type='create' 
      handleAddTeam={handleAddTeam}
      history={history}
      data-test='build-team-component'
    />
  );
};

const mapDispatchToProps = (dispatch) => ({
  addPokeTeam : (pokeTeam) => dispatch(addPokeTeam(pokeTeam))
});

export default connect(null, mapDispatchToProps)(CreateTeamPage);
