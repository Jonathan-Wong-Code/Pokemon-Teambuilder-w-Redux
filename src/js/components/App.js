import React from 'react';
import { Router, Switch } from 'react-router-dom';
import createHistory from 'history/createBrowserHistory';
import CreateTeamPage from './CreateTeamPage';
import SavedTeamsPage from './SavedTeamsPage';
import EditTeamPage from './EditTeamPage';
import LoginPage from './LoginPage';
import PrivateRoute from './../routes/PrivateRoute';
import PublicRoute from './../routes/PublicRoute';

const history = createHistory();

const App = () => {
  return (
    <Router history={history}>
      <React.Fragment>
        <Switch>
          <PublicRoute 
            path='/' 
            component={LoginPage} 
            exact
          />

          <PrivateRoute 
            path='/create' 
            component={CreateTeamPage} 
          />
          <PrivateRoute
            path='/SavedTeams' 
            component={SavedTeamsPage} 
          />
          <PrivateRoute 
            path='/edit/:id' 
            component={EditTeamPage} 
          />
        </Switch>
      </React.Fragment>
    </Router>
  );
};

export default App;
