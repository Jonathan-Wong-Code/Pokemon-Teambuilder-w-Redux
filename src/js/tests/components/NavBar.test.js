import React from 'react';
import { shallow } from 'enzyme';
import { NavBar } from './../../components/NavBar';
import { loginGuest } from '../../actions/auth';

describe("The unconnected NavBar component with non-guest account", () => {
  let wrapper, logout, history, startLogout, auth;
  beforeEach(() => {
    logout = jest.fn();
    history = { push : jest.fn() };
    startLogout = jest.fn();
    auth = 'test123';
    const defaultProps = {
      logout,
      history,
      startLogout,
      auth
    };
    wrapper = shallow(<NavBar {...defaultProps} />);
  });

  it('Should render a snapshot of the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Should render the NavBar component', () => {
    expect(wrapper.find(`[data-test='nav-bar-component']`).length).toBe(1);
  });

  it('Should log the user out when logout clicked', () => {
    wrapper.find(`[data-test='nav-bar-logout']`).simulate('click');
    expect(startLogout).toHaveBeenCalled();
  });

  it('Should log the user out of the guest account and redirected to home screen', () => {
    wrapper.setProps({
      auth : 'guest'
    });

    wrapper.find(`[data-test='nav-bar-logout']`).simulate('click');
    expect(logout).toHaveBeenCalled();
    expect(history.push).toHaveBeenLastCalledWith('/');
  });
});
