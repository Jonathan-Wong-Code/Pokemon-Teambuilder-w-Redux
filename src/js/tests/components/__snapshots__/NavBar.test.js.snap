// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The unconnected NavBar component with non-guest account Should render a snapshot of the component 1`] = `
<nav
  className="navbar"
  data-test="nav-bar-component"
>
  <ul
    className="navbar__nav"
  >
    <li
      className="navbar__item"
    >
      <NavLink
        activeClassName="is-active"
        aria-current="page"
        className="navbar__link btn"
        to="/create"
      >
        Create Team
      </NavLink>
    </li>
    <li
      className="navbar__item"
    >
      <NavLink
        activeClassName="is-active"
        aria-current="page"
        className="navbar__link btn"
        to="/SavedTeams"
      >
        Saved Teams
      </NavLink>
    </li>
    <li
      className="navbar__item"
    >
      <button
        className="btn navbar__logout"
        data-test="nav-bar-logout"
        onClick={[Function]}
      >
        Logout
      </button>
    </li>
  </ul>
</nav>
`;
